ksql-engine/src/main/java/io/confluent/ksql/function/udaf/topk/TopkKudaf.java

```diff
@@ -44,86 +42,65 @@
             final Schema returnType,
             final List<Schema> argumentTypes,
             final Class<T> clazz) {
-    super(functionName,
+    super(
+        functionName,
         argIndexInValue,
-        () -> (T[]) Array.newInstance(clazz, topKSize),
+        ArrayList::new,
         returnType,
         argumentTypes
     );
     this.topKSize = topKSize;
     this.returnType = returnType;
     this.argumentTypes = argumentTypes;
     this.clazz = clazz;
-    this.comparator = (v1, v2) -> {
-      if (v1 == null && v2 == null) {
-        return 0;
-      }
-      if (v1 == null) {
-        return 1;
-      }
-      if (v2 == null) {
-        return -1;
-      }
-
-      return Comparator.<T>reverseOrder().compare(v1, v2);
-    };
   }
 
   @SuppressWarnings("unchecked")
   @Override
-  public T[] aggregate(final T currentValue, final T[] aggregateValue) {
+  public List<T> aggregate(final T currentValue, final List<T> aggregateValue) {
     if (currentValue == null) {
       return aggregateValue;
     }
 
-    final T last = aggregateValue[aggregateValue.length - 1];
-    if (last != null && currentValue.compareTo(last) <= 0) {
-      return aggregateValue;
+    final int currentSize = aggregateValue.size();
+    if (!aggregateValue.isEmpty()) {
+      final T last = aggregateValue.get(currentSize - 1);
+      if (currentValue.compareTo(last) <= 0
+          && currentSize == topKSize) {
+        return aggregateValue;
+      }
     }
 
-    final int nullIndex = ArrayUtil.getNullIndex(aggregateValue);
-    if (nullIndex != -1) {
-      aggregateValue[nullIndex] = currentValue;
-      Arrays.sort(aggregateValue, comparator);
-      return aggregateValue;
+    if (currentSize == topKSize) {
+      aggregateValue.set(currentSize - 1, currentValue);
+    } else {
+      aggregateValue.add(currentValue);
     }
 
-    aggregateValue[aggregateValue.length - 1] = currentValue;
-    Arrays.sort(aggregateValue, comparator);
+    aggregateValue.sort(Comparator.reverseOrder());
     return aggregateValue;
   }
 
   @SuppressWarnings("unchecked")
   @Override
-  public Merger<String, T[]> getMerger() {
-    return (aggKey, aggOne, aggTwo) -> {
-      final T[] merged = (T[]) Array.newInstance(clazz, topKSize);
-
-      int idx1 = 0;
-      int idx2 = 0;
-      for (int i = 0; i != topKSize; ++i) {
-        final T v1 = idx1 < aggOne.length ? aggOne[idx1] : null;
-        final T v2 = idx2 < aggTwo.length ? aggTwo[idx2] : null;
-
-        if (comparator.compare(v1, v2) < 0) {
-          merged[i] = v1;
-          idx1++;
-        } else {
-          merged[i] = v2;
-          idx2++;
-        }
+  public Merger<String, List<T>> getMerger() {
```

Comment:
Previous flagged as a performance concern. Your impl takes over 6 minutes on my machine to run the test in `IntTopkKudfTest`. The impl below, based heavily on the code that was already there, takes around 1:20. So yours update is 3.5 times slower.

```
public Merger<String, List<T>> getMerger() {
    return (aggKey, aggOne, aggTwo) -> {
      final ArrayList<T> merged = new ArrayList<>(Math.min(topKSize, aggOne.size() + aggTwo.size()));

      int idx1 = 0;
      int idx2 = 0;
      for (int i = 0; i != topKSize; ++i) {
        final T v1 = idx1 < aggOne.size() ? aggOne.get(idx1) : null;
        final T v2 = idx2 < aggTwo.size() ? aggTwo.get(idx2) : null;

        if (v1 != null && (v2 == null || v1.compareTo(v2) >= 0)) {
          merged.add(v1);
          idx1++;
        } else if (v2 != null && (v1 == null || v1.compareTo(v2) < 0)) {
          merged.add(v2);
          idx2++;
        } else {
          break;
        }
      }

      return merged;
    };
}
```

Reply:
Changed it to the above code.